lib/schema-controller.js

```diff
@@ -0,0 +1,169 @@
+
+/**
+ * When the dev will execute `fastify.getSchemas` this method will be called.
+ * This should return all the schemas added by the user in the fastify context
+ */
+SchemaController.prototype.getSchemas = function getSchemas () { return this.schemaBucket.getSchemas() }
+
+/**
+ * When the dev will execute `fastify.setValidatorCompiler` this method will be called.
+ * @param {function} validatorCompiler: the validation compiler factory
+ */
+SchemaController.prototype.setValidatorCompiler = function (validatorCompiler) { this.validatorCompiler = validatorCompiler }
+
+/**
+ * When the dev will execute `fastify.setSerializerCompiler` this method will be called.
+ * @param {function} serializerCompiler: the serialization compiler factory
+ */
+SchemaController.prototype.setSerializerCompiler = function (serializerCompiler) { this.serializerCompiler = serializerCompiler }
+
+/**
+ * When the dev will execute `fastify.validatorCompiler` this method will be called.
+ */
+SchemaController.prototype.getValidatorCompiler = function () {
+  return this.validatorCompiler || (this.parent && this.parent.getValidatorCompiler())
+}
+
+/**
+ * When the dev will execute `fastify.serializerCompiler` this method will be called.
+ */
+SchemaController.prototype.getSerializerCompiler = function () {
+  return this.serializerCompiler || (this.parent && this.parent.getSerializerCompiler())
+}
+
+/**
+ * This method will be called when a validator must be setup.
+ * Do not setup the compiler more than once
+ * @param {object} serverOptions: the fastify server option
+ */
+SchemaController.prototype.setupValidator = function (serverOption) {
+  const isReady = this.validatorCompiler !== undefined && !this.schemaBucket.hasNewSchemas()
+  if (isReady) {
+    return
+  }
+  this.validatorCompiler = buildPerformanceValidator(this.schemaBucket.getSchemas(), serverOption.ajv)
+}
+
+/**
+ * This method will be called when a serializer must be setup.
+ * Do not setup the compiler more than once
+ * @param {object} serverOptions: the fastify server option
+ */
+SchemaController.prototype.setupSerializer = function (serverOption) {
+  const isReady = this.serializerCompiler !== undefined && !this.schemaBucket.hasNewSchemas()
+  if (isReady) {
+    return
+  }
+  this.serializerCompiler = buildDefaultSerializer(this.schemaBucket.getSchemas())
+}
+
+module.exports = SchemaController
+
+function buildDefaultSerializer (externalSchemas) {
+  return function ({ schema, method, url, httpStatus }) {
+    return fastJsonStringify(schema, { schema: externalSchemas })
+  }
+}
+
+function ValidatorSelector () {
+  const validatorPool = new Map()
+  const cache = new Map()
+  cache.put = cache.set
+
+  return function buildCompilerFromPool (externalSchemas, options) {
+    const externals = JSON.stringify(externalSchemas)
+    const ajvConfig = JSON.stringify(options.customOptions)
```

Comment:
these two `JSON.stringify()` seem pretty expensive to do. Are they invoked often?

Reply:
They are invoked if:

- the fastify context doesn't have a schema validator (NB: when it is set up for the root instance, it is inherited by children)
- the fastify context has more schemas than its parent

So:
- in the best case it is executed once (when the shared schemas are added at the root fastify context)
- in the worse case for every fastify context

Looking the code the `const ajvConfig = JSON.stringify(options.customOptions)` could be avoided since it is constant overtime and the deferred loading of schema added (your comment below) could avoid this check.